connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java

```diff
@@ -302,7 +326,12 @@ public Future<Void> set(Map<ByteBuffer, ByteBuffer> values, Callback<Void> callb
                 }
             }
             try (LoggingContext context = loggingContext()) {
-                callback.onCompletion(primaryWriteError, ignored);
+                Throwable secondaryWriteError = secondaryStoreTombstoneWriteError.get();
+                if (secondaryStore != null && containsTombstones && secondaryWriteError != null) {
+                    callback.onCompletion(secondaryWriteError, ignored);
+                } else {
+                    callback.onCompletion(primaryWriteError, ignored);
```

Comment:
There's one case here where the write to secondary store is successful but fails to primary store. In such a case, there would be discrepancies in the 2 offsets topic but since that would also lead to the failure of offset commit, that discrepancy is still different from the originally reported issue. Let me know if that makes sense.